File: merge_two_sorted_list/merge_two_sorted_lists.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function mergeTwoLists($l1, $l2) {
        
        if($l1 == null && $l2 == null) return null;
        if($l1 == null && $l2 != null) return $l2;
        if($l1 != null && $l2 == null) return $l1;
        
        if($l1->val <= $l2->val){
            $l3 = $l1;
            $l1 = $l1->next;
        } else {
            $l3 = $l2;
            $l2 = $l2->next;
        }
        
        $result = $l3;

        while($l1 !== null && $l2 !== null){
            if($l1->val <= $l2->val){
                $l3->next = $l1;
                $l1 = $l1->next;
            } else {
                $l3->next = $l2;
                $l2 = $l2->next;
            }
            $l3 = $l3->next;
        }
        
        while(!is_null($l1)){
            $l3->next = $l1;
            $l1 = $l1->next;
            $l3 = $l3->next;
        }
        
        while(!is_null($l2)){
            $l3->next = $l2;
            $l2 = $l2->next;
            $l3 = $l3->next;
        }
        
        return $result;
    }
}

?>

File: merge_two_sorted_list/merge_two_sorted_lists.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// ListNode definition
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	result := &ListNode{}
	l3 := &ListNode{}

	if l1 == nil && l2 == nil {
		return nil
	}

	if l1 == nil && l2 != nil {
		return l2
	}

	if l1 != nil && l2 == nil {
		return l1
	}
	if l1.Val <= l2.Val {
		l3 = l1
		l1 = l1.Next
	} else {
		l3 = l2
		l2 = l2.Next
	}

	result = l3

	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			l3.Next = l1
			l1 = l1.Next
		} else {
			l3.Next = l2
			l2 = l2.Next
		}
		l3 = l3.Next
	}

	for l1 != nil {
		l3.Next = l1
		l1 = l1.Next
		l3 = l3.Next
	}

	for l2 != nil {
		l3.Next = l2
		l2 = l2.Next
		l3 = l3.Next
	}

	return result
}

func main() {}
